---
title: "Parkinson Meta-Analysis"
author: "Sven Kleine Bardenhorst"
date: "8 4 2020"
output:
  rmdformats::material:
    highlight: kate
    thumbnails: true
    code_folding: show
    fig_width: 6
    fig_height: 4
    df_print: paged
---
```{css, echo = FALSE}
.header-panel {
    background-color: #940a21;
    min-height: 144px;
    position: relative;
    z-index: 3;
}
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE,fig.height=5, fig.width=10)
```


```{r load packages, message=FALSE, warning=FALSE, include=FALSE}
#### load packages ####
source('packages.R')

#### load functions ####
source("functions.R")
```

# Preparation of data based on same sequence region (V4)

### Load data
```{r Load data, message=FALSE, warning=FALSE}
# Load ASV data
A_data <- readr::read_delim("Data/A_data.txt",delim="\t")
B_data <- readr::read_delim("Data/B_data.txt",delim="\t")
C_data <- readr::read_delim("Data/C_data.txt",delim="\t")
E_data <- readr::read_delim("Data/E_data.txt",delim="\t")
F_data <- readr::read_delim("Data/F_data.txt",delim="\t")
H_data <- readr::read_delim("Data/H_data.txt",delim="\t")
I_data <- readr::read_delim("Data/I_data.txt",delim="\t")
J_data <- readr::read_delim("Data/J_data.txt",delim="\t")
```


### Transform into tidy format
```{r}
# transpose into tidy format
A_dat <- tidy_up(A_data,trim=F)
B_dat <- tidy_up(B_data,trim=T)
C_dat <- tidy_up(C_data)
E_dat <- tidy_up(E_data,trim=T)
F_dat <- tidy_up(F_data)
H_dat <- tidy_up(H_data)
I_dat <- tidy_up(I_data)  #%>% mutate(ID = str_replace_all(ID,"-","."))
J_dat <- tidy_up(J_data)
```


### Prepare taxonomy based on V4-region trimmed sequences sequences
```{r echo=TRUE, message=FALSE, warning=FALSE}
studies <- list("A_Dysbiosisofgutmicrobiota",
             "B_Alterationofthefecalmicrobiota_2",
             "C_ColonicbacterialcompositioninParkinson",
             "E_Gutmicrobiotain",
             "F_ThenasalandgutmicrobiomeinParkinsonsdisease",
             "H_EffectofParkinsonsdisease",
             "I_UnravellingGutMicrobiotas")

author.names <- c("Pietrucci et al., 2019",
           "Qian et al., 2018",
           "Keshavarzian et al., 2015",
           "Aho et al., 2019",
           "Heintz-Buschart et al., 2018",
           "Weis et al., 2019",
           "Barichella et al., 2019")

#project_path <- c("G:\\Projects\\ParkinsonMicrobiome\\")
project_path <- c("C:\\Users\\kleinebardenhorst\\Documents\\Projects\\ParkinsonMicrobiome\\ParkinsonMicrobiome\\")



path <- paste0(project_path,"Data\\")
seq_same <- paste("Sequences_sameregion\\")
tax_same <- paste("Classified_sameregion\\Classified_Unaligned_Sliced_F_Sequences_") 

taxa <- lapply(studies,function(x) read_tsv(paste0(path,tax_same,x,"_noC.txt.align.fasta"),quote="",col_names = F))
seq <- lapply(studies, function(x) ((seqinr::read.fasta(paste0(path,seq_same,"Unaligned_Sliced_F_Sequences_",x,"_noC.txt.align.fasta"),as.string = T,set.attributes = F))))

seq <- foreach(i=seq_along(seq)) %do% data.frame("ID"=paste0("ASV",names(unlist(seq[[i]]))),"Sequence"=unlist(seq[[i]]))


taxon <- lapply(taxa,prepare_taxonomy_sameregion)

A_tax <- taxon[[1]] %>% left_join(seq[[1]]) %>% apply(2,function(x) gsub('\"', "", x, fixed = TRUE)) %>% as_tibble()
B_tax <- taxon[[2]] %>% left_join(seq[[2]])  %>% apply(2,function(x) gsub('\"', "", x, fixed = TRUE)) %>% as_tibble()
C_tax <- taxon[[3]] %>% left_join(seq[[3]]) %>% apply(2,function(x) gsub('\"', "", x, fixed = TRUE)) %>% as_tibble()
E_tax <- taxon[[4]] %>% left_join(seq[[4]]) %>% apply(2,function(x) gsub('\"', "", x, fixed = TRUE)) %>% as_tibble()
F_tax <- taxon[[5]] %>% left_join(seq[[5]]) %>% apply(2,function(x) gsub('\"', "", x, fixed = TRUE)) %>% as_tibble()
H_tax <- taxon[[6]] %>% left_join(seq[[6]]) %>% apply(2,function(x) gsub('\"', "", x, fixed = TRUE)) %>% as_tibble()
I_tax <- taxon[[7]] %>% left_join(seq[[7]]) %>% apply(2,function(x) gsub('\"', "", x, fixed = TRUE)) %>% as_tibble()


```

### Quality filtering

```{r}
A_clean <- clean(A_dat,A_tax,filter="rel")
B_clean <- clean(B_dat,B_tax,filter="rel")
C_clean <- clean(C_dat,C_tax,filter="rel")
E_clean <- clean(E_dat,E_tax,filter="rel")
F_clean <- clean(F_dat,F_tax,filter="rel")
H_clean <- clean(H_dat,H_tax,filter="rel")
I_clean <- clean(I_dat,I_tax,filter="rel")


#ind <- I_tax[I_tax$Phylum=="Bacteroidetes","ID"] %>% pull

#apply(I_dat[,ind],1,sum)

#I_clean[,names(I_clean) %in% ind]
```


### Prepare metadata
```{r message=FALSE, warning=FALSE}
# Load ENA metadata
A_Meta_ENA <- loadENAMeta(path="Data/A_meta_ENA.txt") %>% dplyr::select(run_accession,library_name)
B_Meta_ENA <- loadENAMeta(path="Data/B_meta_ENA.txt")
#C_Meta <- loadENAMeta(path="Data/C_meta.txt")
#D_Meta <- loadENAMeta(path="Data/D_meta.txt")
E_Meta_ENA <- loadENAMeta(path="Data/E_meta_ENA.txt") %>% 
  dplyr::select(run_accession,sample_title) %>%
  slice(1:266) %>% # Slice blank controls
  mutate(Timepoint = str_split(.$sample_title," ",simplify = TRUE)[,2],
         Subject = str_split(.$sample_title," ",simplify = TRUE)[,1]) %>%
  dplyr::select(-sample_title)

F_Meta_ENA <- loadENAMeta(path="Data/F_meta_ENA.txt") %>% dplyr::select(run_accession, library_name,experiment_alias)
G_Meta_ENA <- loadENAMeta(path="Data/G_meta_ENA.txt")
H_Meta_ENA <- loadENAMeta(path="Data/H_meta_ENA.txt") %>%
  dplyr::select(run_accession,sample_alias) %>%
  slice(1:256) %>% # Slice ion torrent
  mutate(ID = run_accession,
    Sample = str_split(.$sample_alias,"_",simplify = T)[,1]) %>%
  dplyr::select(ID,Sample)


#I_Meta <- loadENAMeta(path="Data/I_meta.txt")
J_Meta_ENA <- loadENAMeta(path="Data/J_meta_ENA.txt")

# Load detailed metadata
project_path

C_meta1 <- xlsx::read.xlsx(paste0(project_path,"Data/Meta/C_meta.xlsx"),sheetIndex = 1, header = T,startRow =2)
names <- str_remove(colnames(C_meta1),"PD.")
colnames(C_meta1) <- names

C_meta2 <- xlsx::read.xlsx(paste0(project_path,"Data/Meta/C_meta.xlsx"),sheetIndex = 2, header = T,startRow =2)
names <- str_remove(colnames(C_meta2),"HC.")
colnames(C_meta2) <- names

C_Meta <- C_meta1 %>% full_join(C_meta2) %>% filter(Feces %in% toupper(C_dat$ID) | Tissue %in% toupper(C_dat$ID))

E_meta <- readr::read_csv(paste0(project_path,"Data/Meta/E_meta.csv")) %>% 
  left_join(E_Meta_ENA) %>% dplyr::select(run_accession,everything())

H_meta <- xlsx::read.xlsx(paste0(project_path,"Data/Meta/H_meta.xlsx"),
                          sheetIndex = 1, header = T,startRow =1) # Other drugs variable not used and deleted for easier import


I_meta <- xlsx::read.xlsx(paste0(project_path,"Data/Meta/I_meta.xls"), sheetIndex = 1,header=T,startRow = 2) %>% slice(1:306)

J_meta <- readr::read_csv(paste0(project_path,"Data/Meta/J_meta.csv")) %>%
  dplyr::select(cc,ID,Age,Sex,Batch,CC,Onset,'UPDRS-I','UPDRS-III',Stool_freq,Constipation,Vegetarian,Nicotine,Alcohol,Caffeine,Ferm_Milk,Age_DBS,L_Dopa,Benserazid,Carbidopa,Dopa_Agon,COMT_Inhibitor,MAO_B_Inh,CV_disease)
```

### Prepare phyloseq objects

```{r echo=FALSE, message=FALSE, warning=FALSE}

# Combine Taxonomies, Metadata and ASV-Data into phyloseq objects

# Study A
## OTU
OTU_A <- as.matrix(A_clean[,-1])
rownames(OTU_A) <- A_dat$ID
OTU_A <- otu_table(t(OTU_A),taxa_are_rows = T)

## Tax
tax_A <- as.matrix(A_tax[,-1])
rownames(tax_A) <- A_tax$ID

tax_A <- tax_table(tax_A)

## Meta
A_sample <- A_Meta_ENA %>%
            mutate(pd = substr(.$library_name,1,3),
                   author = "Pietrucci et al., 2019") %>%
            mutate(status = ifelse(pd=="CON","HC","PD")) %>%
            column_to_rownames(var="run_accession") %>% sample_data()

## Combine
A_phyl <- phyloseq(OTU_A,tax_A,A_sample)


# Study B

## OTU
OTU_B <- as.matrix(B_clean[,-1])
rownames(OTU_B) <- B_dat$ID
OTU_B <- otu_table(t(OTU_B),taxa_are_rows = T)

## Tax
tax_B <- as.matrix(B_tax[,-1])
rownames(tax_B) <- B_tax$ID
tax_B <- tax_table(tax_B)

## Meta
B_sample <- B_Meta_ENA %>%
  dplyr::select(run_accession,library_name) %>%
  mutate(status = case_when(startsWith(library_name,"GC") ~ "HC",
                            startsWith(library_name,"GP") ~ "PD"),
         author = "Qian et al., 2018") %>%
  filter(not.na(status)) %>% # Filter out blood metagenome
  column_to_rownames(var="run_accession") %>%
  sample_data()

## Combine
B_phyl <- phyloseq(OTU_B,tax_B,B_sample)


# Study C
## OTU
OTU_C <- as.matrix(C_clean[,-1])
rownames(OTU_C) <- toupper(C_dat$ID)
OTU_C <- otu_table(t(OTU_C),taxa_are_rows = T)

## Tax
tax_C <- as.matrix(C_tax[,-1])
rownames(tax_C) <- toupper(C_tax$ID)

tax_C <- tax_table(tax_C)

## Meta
C_sample <- C_Meta %>%
  dplyr::select(Case.IDs,Tissue,Feces,everything()) %>%
  tidyr::gather(Sample,ID,2:3) %>%
  dplyr::select(Case.IDs,Sample,ID,everything()) %>%
  filter(ID != 'NA') %>%
  filter(Sample=="Feces") %>%
  mutate(status = ifelse(startsWith(Cases,"Parkinson"),"PD","HC"),
         author = "Keshavarzian et al., 2015") %>%
  dplyr::select(status,everything()) %>%
  column_to_rownames(var='ID') %>%
  sample_data

## Combine
C_phyl <- phyloseq(OTU_C,tax_C,C_sample)


# Study E
## OTU
OTU_E <- as.matrix(E_clean[,-1])
rownames(OTU_E) <- E_dat$ID
OTU_E <- otu_table(t(OTU_E),taxa_are_rows = T)

## Tax
tax_E <- as.matrix(E_tax[,-1])
rownames(tax_E) <- E_tax$ID

tax_E <- tax_table(tax_E)

## Meta
E_sample <- E_meta %>%
            filter(Timepoint == "followup") %>%
            mutate(status = ifelse(Parkinson=="control","HC","PD"),
                   author = "Aho et al., 2019") %>%
            dplyr::select(status,everything()) %>%
            column_to_rownames(var='run_accession') %>%
            sample_data()

## Combine
E_phyl <- phyloseq(OTU_E,tax_E,E_sample)


# Study F
## OTU
OTU_F <- as.matrix(F_clean[,-1])
rownames(OTU_F) <- F_dat$ID
OTU_F <- otu_table(t(OTU_F),taxa_are_rows = T)

## Tax
tax_F <- as.matrix(F_tax[,-1])
rownames(tax_F) <- F_tax$ID

tax_F <- tax_table(tax_F)

## Meta
F_sample <- F_Meta_ENA %>%
  filter(endsWith(library_name,".f")) %>%
  filter(startsWith(library_name,"PD")|startsWith(library_name,"HC")) %>%
  mutate(status = ifelse(startsWith(library_name,"PD"),"PD","HC"),
         author = "Heintz-Buschart et al., 2018") %>%
  column_to_rownames(var="run_accession") %>% sample_data


## Combine
F_phyl <- phyloseq(OTU_F,tax_F,F_sample)


# Study H
## OTU
OTU_H <- H_clean %>%
  left_join(H_Meta_ENA) %>%
  group_by(Sample) %>%
  summarize_if(is.numeric,funs(sum)) %>%
  arrange(desc(Sample))

id <- substr(OTU_H$Sample,4,8)
id[40:length(id)] <- paste0("P",id[40:length(id)])

OTU_H$Sample[1:39] <-  paste0('IfM ',substr(id[1:39],1,1),"0",substr(id[1:39],2,4))
OTU_H$Sample[40:nrow(OTU_H)] <- paste0('IfM ',substr(id[40:length(id)],1,2),"0",substr(id[40:length(id)],3,5))

OTU_H <- OTU_H %>% column_to_rownames(var="Sample") %>% as.matrix %>% t %>% otu_table(taxa_are_rows = T)
## Tax
tax_H <- as.matrix(H_tax[,-1])
rownames(tax_H) <- H_tax$ID
tax_H <- tax_table(tax_H)

## Meta
H_sample <- H_meta %>%
            mutate(status = ifelse(Group =="PD","PD","HC"),
                   author = "Weis et al., 2019") %>%
            dplyr::select(status,everything(),-Group) %>%
            column_to_rownames(var="SampleID") %>%
            sample_data

## Combine
H_phyl <- phyloseq(OTU_H,tax_H,H_sample)


# Study I

## OTU
OTU_I <- as.matrix(I_clean[,-1])
rownames(OTU_I) <- str_replace_all(I_dat$ID,"-",".")
OTU_I <- otu_table(t(OTU_I),taxa_are_rows = T)

## Tax
tax_I <- as.matrix(I_tax[,-1])
rownames(tax_I) <- I_tax$ID

tax_I <- tax_table(tax_I)

## Meta
I_sample <- I_meta %>%
  mutate(status = ifelse(status==0,"HC","PD"),
         author = "Barichella et al., 2019")  %>%
            column_to_rownames(var="ID") %>%
            sample_data


## Combine
I_phyl <- phyloseq(OTU_I,tax_I,I_sample)


A_phyl
B_phyl
C_phyl
E_phyl
F_phyl
H_phyl
I_phyl

```

## Filter samples with library size below 10000 reads
```{r}
A_phyl <- prune_samples(sample_sums(A_phyl)>=10000, A_phyl)
B_phyl <- prune_samples(sample_sums(B_phyl)>=10000, B_phyl) ## 19 Samples filtered
C_phyl <- prune_samples(sample_sums(C_phyl)>=10000, C_phyl) ## 2 samples filtered
E_phyl <- prune_samples(sample_sums(E_phyl)>=10000, E_phyl) ## 2 Samples filtered
F_phyl <- prune_samples(sample_sums(F_phyl)>=10000, F_phyl) ## 1 sample filtered
H_phyl <- prune_samples(sample_sums(H_phyl)>=10000, H_phyl)
I_phyl <- prune_samples(sample_sums(I_phyl)>=10000, I_phyl) ## 47 samples filtered

```




# Data inspection

## Number of taxa per phylum
```{r}
dat <- list(A_phyl,
            B_phyl,
            C_phyl,
            E_phyl,
            F_phyl,
            H_phyl,
            I_phyl)


fpp <- plyr::ldply(sapply(dat,features_per_phylum),rbind)
rownames(fpp) <- author.names
fpp
```


## Prevelance and abundance of taxa per phylum
We can have a look at the prevalence of the phyla to see whether a phylum is dominated by a single taxon or by many different taxa. Prevelance here means, whether a specific taxon is present in a subject. As example, the total prevalence of Acidobateria in **Pietrucci et al., 2019** means, that - in total - the 4 taxa (as identified in table above) of the phylum Acidobacteria appeared 55 times. On average, each taxon appeared 13.75 times, indicating that the prevalence of the taxa are relatively equally distributed within the phylum Acidobacteria.
The tables should always be interpreted in combination with the table above (and the figures for abundance information).


### Pietrucci et al., 2019
```{r, fig.cap = "Pietrucci et al., 2019"}
prevalence_per_phylum(A_phyl) %>% kable(caption=paste0(author.names[[1]])) %>% kable_styling()
prevalence_by_abundance(A_phyl,thrs = 0)
```
* Deinococcus-Thermus
  + only one taxon
  + appears in only one subject
* Spirochaetes
  + only one taxon
  + appears in only one subject
* Elusimicrobia !?
  + only 5 taxa
  + appears in only 2.5% of cases on average
  + abundance between 300-1000

### Qian et al., 2018
```{r, fig.cap = "Qian et al., 2018"}
prevalence_per_phylum(B_phyl) %>% kable(caption=paste0(author.names[[2]])) %>% kable_styling()
prevalence_by_abundance(B_phyl,thrs = 0)
```

* candidate division WPS-2 
  + only one taxon
  + appears in one subject
  + low abundance < 150
* Deferribacteres
  + low prevalence
  + appear in two subjects
  + low abundance < 150
* Deinococcus-Thermus
  + low prevalence
  + one subject
  + low abundance < 300
* Planctomycetes
  + one taxon
  + one subject
  + low abundance < 150
* Tenericutes
  + one taxon
  + one subject
  + but, HIGH abundance > 16000


### Keshavarzian et al., 2015
```{r, fig.cap = "Keshavarzian et al., 2015"}
prevalence_per_phylum(C_phyl) %>% kable(caption=paste0(author.names[[3]])) %>% kable_styling()
prevalence_by_abundance(C_phyl,thrs = 0)
```

* Planctomycetes
* Spirochaetes
* Tenericutes
  + low prevalence
  + one subject on average
  + abundance ~ 1300
* Fusobacteria
  + low prevalence
  + one subject on average
  + suber low abundance < 25

### Aho et al., 2019
```{r, fig.cap = "Aho et al., 2019"}
prevalence_per_phylum(E_phyl) %>% kable(caption=paste0(author.names[[4]])) %>% kable_styling()
prevalence_by_abundance(E_phyl,thrs = 0)
```

* Deferribacteres
  + one taxon
  + six subjects
  + low abundance ~ 200
* Elusimicrobia
  + one taxon
  + one subject
  + abundance ~ 500
* Fusobacteria
  + 3 taxa
  + low prevalence < 1
  + extreme range in abundance - 300 < abundance < 10000

### Heintz-Buschart et al., 2018
```{r, fig.cap = "Heintz-Buschart et al., 2018"}
prevalence_per_phylum(F_phyl) %>% kable(caption=paste0(author.names[[5]])) %>% kable_styling()
prevalence_by_abundance(F_phyl,thrs = 0)
```
* Armatimonadetes
* candidate division WPS-1
* candidate division WPS-2
* Chloroflexi
* Chrysiogenetes
* Deferribacteres
* Deinococcus-Thermus
* Gemmatimonadetes
* Hydrogenedentes
* Ignavibacteriae
* Planctomycetes
* Spirochaetes
* Elusimicrobia
  + one taxon
  + three subjects
  + abundance ~ 2000

### Weis et al., 2019
```{r, fig.cap = "Weis et al., 2019"}
prevalence_per_phylum(H_phyl) %>% kable(caption=paste0(author.names[[6]])) %>% kable_styling()
prevalence_by_abundance(H_phyl,thrs = 0)
```

* Chloroflexi
  + one taxon
  + one subject
  + low abundance < 200
* Tenericutes
  + three taxa
  + two subjects
  + low abundance < 350



### Barichella et al., 2019
```{r, fig.cap = "Barichella et al., 2019"}
prevalence_per_phylum(I_phyl) %>% kable(caption=paste0(author.names[[7]])) %>% kable_styling()
prevalence_by_abundance(I_phyl,thrs = 0)
```
* Deferribacteres
  + one taxon
  + three subjects
  + low abundance < 400
* Elusimicrobia
  + one taxon
  + two subjects
  + abundance ~ 1600
* Tenericutes
  + one taxon
  + one subject
  + abundance ~ 4000

# Pooled analysis of alpha diversity


## Alpha disversity analysis by rarefaction and interpolation of Hill Numbers
```{r}
A_next <- iNEXT::iNEXT(data.frame(t(A_dat[,-1])),q=0)
p <- ggiNEXT(A_next,type = 1) +
  theme(legend.position = "none")


p.dat <- fortify(A_next,type=1)

p.dat$method <- factor(p.dat$method, c("interpolated", "extrapolated"),c("interpolation", "extrapolation"))

ggplot(aes(x=x,y=y,fill=site,col=site),data=p.dat) +
  geom_line(aes(linetype=method),size=1,alpha=.4) +
  theme(legend.position = "none")

```



## Prepare alpha diversity measures
```{r echo=TRUE, message=FALSE, warning=FALSE}
dat <- list(A_phyl,
            B_phyl,
            C_phyl,
            E_phyl,
            F_phyl,
            H_phyl,
            I_phyl)

alpha_dat <- estimate_alpha(dat)
```


```{r echo=TRUE, fig.height=7, fig.width=10, message=FALSE, warning=FALSE}

ggplot(aes(x=as.factor(author),y=(richness)),data=alpha_dat) +
  geom_boxplot(aes(fill=status)) +
  scale_fill_brewer(palette = "Set1") +
  theme(axis.text.x = element_text(angle = 90))

ggplot(aes(x=as.factor(author),y=(shannon)),data=alpha_dat) +
  geom_boxplot(aes(fill=status)) +
  scale_fill_brewer(palette = "Set1") +
  theme(axis.text.x = element_text(angle = 90))

ggplot(aes(x=as.factor(author),y=inv.simpson),data=alpha_dat) +
  geom_boxplot(aes(fill=status)) +
   scale_fill_brewer(palette = "Set1") +
  theme(axis.text.x = element_text(angle = 90))
```

## Estimate alpha diversity by effective number of taxa through hill numbers

The problems with indices like Simpson or Shannon is that they are not diversities by themselves, but only proxies (indices) of the true underlying diversity. However, often every index is treated like a true diversity, which may lead to misleading results for several reasons. At first, indices define diversity in different ways, e.g. some focus solely on the number of present taxa (further referred to as richness), while others incorporate the relative abundance of all present taxa (further referred to as eveness). The combination of speciess richness and eveness into one index makes it hard to interpret. It further leads to the second problem, that some indices do NOT double when the true diversity doubles, hence, they are highly non-linear. As an example, a community with a Shannon-Index ($H'$) of 2 is not twice as diverse as a community with a $H'$ of 1. In addition, on the $H'$-scale, the difference between 1 and 2 is not equal to the difference between 3 and 4. Thus, the location of a mean difference on the $H'$-scale matters. This makes the statisical analysis and interpretation of alpha diversity indices problematic. However, there is a straightforward and intuitive way to define a true diversity in terms of **effective number of species** through hill numbers.

```{r}
x <- rep(150,50)
shannon <- function(x) {-sum( (x/sum(x)) * log(x/sum(x)) )}
s_x <- shannon(x)
```

Given a specific value of $H'$, we cannot derive the true diversity of the community, as we cannot decompose the value into its richness and eveness parts. However, we still know that two communities with an equal index are always equally diverse. 

Suppose we have two communities with `r length(x)` equally abundant taxa. We can calculate the Shannon-Index for this community with

$$
H' = - \sum_i p_i * ln  p_i
$$


with $p_i = \frac{n_i}{N}$. For the given example, $H'$ is `r s_x`. Now, we know the $H'$ of a community with `r length(x)` equally abundant taxa. Therefore, we can conclude that every community with an $H'$ of `r s_x` is always as diverse as a community with `r length(x)` equally abundant taxa.

We can make use of this fact, by transforming the $H'$ into the **effective number of species**, which is simply done by exponantation of $H'$.

$$
\exp(H') = exp(3.912023) = 50
$$
The same logic can be applied to other measures of diversity. In fact, the effective number of species based on the most commonly used diversity indices can be calculated directly by the generalized formula

$$
qD = (\sum_{i_=1}^Sp_i^q)^{1/(1-q)},
$$
which is the Hill number of order q. The expression if fully dependent on the choice of q, the exponent of taxa abundances, thus, determining the weight that is assigned to the evenness of taxa. The higher the value of q, the higher the influence of eveness on the value of the index.

Different values of q correspond to the most frequently used diversity measures. Here, we will focus on $q=0$ (species richness), $q=1$ (exponential of the Shannon index; note that it is not exactly 1, as the formula is not defined for $q=1$, but its limit when q tends to 1) and $q=2$ (the inverse of Simpson's concentration index).

### Alpha diveristy analysis through hill numbers

A common problem with all diversity indices is their dependency on the sampling depth (library size). We can inspect this dependency by correlating the value of the hill numbers of different order with the respective library size.

```{r echo=TRUE, message=FALSE, warning=FALSE, results='asis'}
## Asses whether alpha-diversity associated with sequencing depth
alpha_dat %>% group_by(author) %>% summarize(Richness = round(cor(richness,N),2),
                                             Shannon = round(cor(shannon,N),2),
                                             "Inverse simpson" = round(cor(inv.simpson,N),2)) %>% column_to_rownames("author") %>% t %>% kable(caption = "Correlation of alpha diversity indices with library size") %>% kableExtra::kable_styling("striped")
```

In our setting, this challenge is even more complex, as the library size is not only different between individuals, but also cluster within studies and sometimes within study groups nested in these studies.

```{r echo=TRUE, fig.height=7, fig.width=10, message=FALSE, warning=FALSE, fig.cap="Average library size per study and per study group."}
#alpha_dat %>% group_by(author,status) %>% summarize(avg_LS = mean(N)) %>%
#  ggplot(aes(x=author,y=avg_LS,fill=status)) +
#  geom_bar(stat="identity",position = "dodge") +
#   scale_fill_brewer(palette="Set1") +
#  scale_x_discrete(labels=addline_format(unique(alpha_dat$author))) +
#  theme_sjplot()

alpha_dat %>%
  ggplot(aes(x=author,y=N,fill=status)) +
  geom_boxplot() +
   scale_fill_brewer(palette="Set1") +
  scale_x_discrete(labels=addline_format(unique(alpha_dat$author))) +
  theme_sjplot() +
  ylab("Library size") +
  xlab("Study") +
  labs(fill="Disease status")
  
```

```{r echo=TRUE, fig.height=7, fig.width=10, message=FALSE, warning=FALSE, fig.cap="Distribution of library size per study and per study group."}
alpha_dat %>% dplyr::select(status,N,study) %>%
  ggplot(aes(x=N,fill=status)) +
  geom_density(alpha=.6) +
  scale_fill_brewer(palette="Set1") +
  facet_wrap(~alpha_dat$author,ncol = 3,scales = "free_x") +
  theme_sjplot() +
  xlab("Library size") +
  labs(fill="Disease status")
```

A common approach to handle unequal library sizes is by rarefaction, thus, sub-sampling from each sample to achieve equal library sizes that allow for fair comparisons However, this approach is highly criticized for several reasons, among which is the discarding of valuable data. Therefore, we will use a poisson model including the library size as an offset, conditioning the estimates on the library size. Using this approach, we will incorporate the uncertainty associated with the library size (cases will be weighted according to their library size), without throwing away any information.



## Random Effects models for alpha diversity

## Two-stage approach

```{r fig.cap="Two stage meta-analysis using poisson regression with library size as offset - richness."}
alpha_dat2 <- alpha_dat %>% dplyr::select(ID,status,everything()) %>% filter(N>10)

## metafor two-stage approach
library(metafor)

met_dat <- alpha_dat2 %>%
  dplyr::select(study,status,richness,N,author) %>%
  group_by(study,status) %>%
  summarise(richness = sum(richness),
            N = sum(N),
            author = first(author)) %>%
  pivot_wider(names_from=status,values_from=c(richness,N))

res <- rma(measure = "IRR",x2i=richness_HC,x1i=richness_PD,t2i=N_HC,t1i=N_PD,data=met_dat)

forest(res,transf = exp,refline = 1,slab = unique(alpha_dat2$author))
```

```{r fig.cap="Two stage meta-analysis using poisson regression with library size as offset - shannon."}
## metafor two-stage approach

met_dat <- alpha_dat2 %>%
  dplyr::select(study,status,shannon,N,author) %>%
  group_by(study,status) %>%
  summarise(shannon = sum(shannon),
            N = sum(N),
            author = first(author)) %>%
  pivot_wider(names_from=status,values_from=c(shannon,N))

res <- rma(measure = "IRR",x2i=shannon_HC,x1i=shannon_PD,t2i=N_HC,t1i=N_PD,data=met_dat)

forest(res,transf = exp,refline = 1,slab = unique(alpha_dat2$author))
```

```{r fig.cap="Two stage meta-analysis using poisson regression with library size as offset - inv.simpson"}
## metafor two-stage approach
met_dat <- alpha_dat2 %>%
  dplyr::select(study,status,inv.simpson,N,author) %>%
  group_by(study,status) %>%
  summarise(inv.simpson = sum(inv.simpson),
            N = sum(N),
            author = first(author)) %>%
  pivot_wider(names_from=status,values_from=c(inv.simpson,N))

res <- rma(measure = "IRR",x2i=inv.simpson_HC,x1i=inv.simpson_PD,t2i=N_HC,t1i=N_PD,data=met_dat)

forest(res,transf = exp,refline = 1,slab = unique(alpha_dat2$author))
```
## One-stage approach

```{r}
## One-stage approach

## Asses ICC
ICC(outcome = "richness",group="study",data=alpha_dat)
fit <- lme4::glmer(richness~1+(status|study),data=alpha_dat2,family=poisson(link="log"),offset = log(N))
fit2 <- lme4::glmer(shannon~1+(status|study),data=alpha_dat2,family=poisson(link="log"),offset = log(N))
fit3 <- lme4::glmer(inv.simpson~1+(status|study),data=alpha_dat2,family=poisson(link="log"),offset = log(N))
names <- alpha_dat2 %>% group_by(study) %>% slice(1) %>% dplyr::select(author)

exp(ranef(fit)$study) %>% t() %>% kable(caption = "Random effects poisson regression with library size as offset - richness") %>% kableExtra::kable_styling()
exp(ranef(fit2)$study) %>% t() %>% kable(caption = "Random effects poisson regression with library size as offset - shannon") %>% kableExtra::kable_styling()
exp(ranef(fit3)$study) %>% t() %>% kable(caption = "Random effects poisson regression with library size as offset - inv.simpson") %>% kableExtra::kable_styling()
```

## Sensitivity to library size

```{r}
range <- alpha_dat %>% filter(study == 2 & N > 0) %>% dplyr::select(N) %>% range
```


There are some studies in which the library size is highly variable, e.g. the study by *Qian et al. (2018)* ranges from `r range[1] ` to `r range[2]` reads per sample. As samples with low reads may be very unreliable/uninformative, we may decide on a cut-off value for a minimum library size. To make a more informed choice, we may conduct a sensitivity analysis to get a feeling for the influence of the library size on the results.
For this sensitivity analysis, we will perform the analysis repeatedly for different minimum library sizes and plot the coefficients as a function of library size.


```{r fig.height=5, fig.width=10, fig.cap="Sensitivity to library size - richness."}
sensitivity_ls(alpha_dat2, index = "richness")
```

```{r fig.height=5, fig.width=10, fig.cap="Sensitivity to library size - shannon."}
sensitivity_ls(alpha_dat2, index = "shannon")
```

```{r fig.height=5, fig.width=10, fig.cap="Sensitivity to library size - inv.simpson."}
sensitivity_ls(alpha_dat2, index = "inv.simpson")
```

# Beta diversity 

## Prepare full distance matrix

```{r echo=FALSE, message=FALSE}
dat <- list(A_phyl,
            B_phyl,
            C_phyl,
            E_phyl,
            F_phyl,
            H_phyl,
            I_phyl)


beta <- prepare_beta_data(dat,tax="Phylum")

cmdscale(beta$dist,k=2) %>%
  as.data.frame() %>%
  rownames_to_column("ID") %>%
  dplyr::left_join(beta$meta) %>%
  ggplot(aes(x=V1,y=V2,col=author,shape=status)) +
  geom_point()

adonis(beta$dist~status,data = beta$meta,strata = beta$meta$author,permutations = 1000)


plot(betadisper(beta$dist,beta$meta$author))

```

# Dirichlet multinomial mixture clustering
```{r}
dat <- list(A_phyl,
            B_phyl,
            C_phyl,
            E_phyl,
            F_phyl,
            H_phyl,
            I_phyl)

## Prepare data on Phylum level

full.dat <- prepare_DA_data(dat,tax="Phylum")
full.dat <- full.dat %>% dplyr::select(ID,status,author,N,everything()) ## Arrange data in new order


library(DirichletMultinomial)

tax.counts <- full.dat %>%
  dplyr::select(-ID,-status,-author,-N) %>%
  as.matrix
rownames(tax.counts) <- full.dat$ID

#nc <- detectCores()
#cl <- makeCluster(nc-2)
#dm.fit <- parLapply(cl=cl,1:12,dmn,count=tax.counts,seed=1406)
#saveRDS(dm.fit,file="Data/Models/dm.fit")
#stopCluster(cl)

dm.fit <- readRDS("Data/Models/dm.fit")

lplc <- sapply(dm.fit,laplace)

plot(lplc, type="b", xlab="Number of Dirichlet Components", ylab="Model Fit")

best.fit <- dm.fit[[which.min(lplc)]]

splom(fitted(best.fit))

p0 <- fitted(dm.fit[[1]], scale=TRUE)# scale by theta>
p8 <- fitted(best.fit, scale=TRUE)
colnames(p8) <- paste("m", 1:8, sep="")

(meandiff <- colSums(abs(p8 - as.vector(p0))))
sum(meandiff)


diff <- rowSums(abs(p8 - as.vector(p0)))
o <- order(diff, decreasing=TRUE)
cdiff <- cumsum(diff[o]) / sum(diff)
df <- head(cbind(Mean=p0[o], p8[o,], diff=diff[o], cdiff), 10)
df %>% kable %>% kable_styling()

heatmapdmn(tax.counts,fit1=dm.fit[[1]], fitN=best.fit, ntaxa=15)
```

# Firmicuts / Bacteroidetes ratio
```{r}
FB.dat <- full.dat %>%
  mutate(Firmicutes = Firmicutes+1,
         Bacteroidetes = Bacteroidetes +1,
         FB_ratio = ((Firmicutes)/Bacteroidetes)) %>%
  dplyr::select(ID,status,author,N,FB_ratio,Firmicutes,Bacteroidetes)

FB.dat %>% group_by(status) %>% summarize(FB = median(FB_ratio))


ggplot(aes(x=author,y=FB_ratio,fill=status),data=FB.dat) +
  geom_boxplot() +
  scale_x_discrete(labels=addline_format(unique(FB.dat$author))) +
  scale_y_log10()
  
fit.fb <- glmer(as.factor(status) ~ scale(FB_ratio) + (1|author),data=FB.dat,family = binomial,control = glmerControl(optimizer = "bobyqa"),
    nAGQ = 10)

summary(fit.fb)

round(FB.dat$FB_ratio,4)
```

# Network analysis

```{r eval=FALSE, include=FALSE}

A_phyl

n.model <- SpiecEasi::spiec.easi(A_phyl,method="glasso",lambda.min.ratio=1e-2,
                           nlambda=20, pulsar.params=list(rep.num=50))

```


# Differential abundance - Phylum level

## Univariate negative binomial mixed models

```{r message=FALSE, warning=FALSE}
taxa <- names(full.dat)[-c(1:4)] # Extract all taxa names

univ.fit <- sapply(taxa,fitZINBMM,simplify = F)

comp.error <- taxa[sapply(univ.fit,is.null)]

#fit.sig <- univ.fit[unlist(lapply(univ.fit[!sapply(univ.fit,is.null)],is.sig))]

results.phyl <- round(t(sapply(univ.fit[!sapply(univ.fit,is.null)],summarize_univariate)),5) %>%
  data.frame() %>%
  rownames_to_column("taxon") %>%
  mutate(FDR.adjusted = p.adjust(p.value,method="fdr")) %>%
  dplyr::select(taxon,estimate,p.value,FDR.adjusted,theta) %>% 
  arrange(FDR.adjusted)

results.phyl %>% kable() %>% kable_styling()

```

```{r}
ggplot(data=full.dat,aes(x=author,y=Bacteroidetes,fill=status)) +
  geom_violin() + 
  geom_jitter() +
  scale_x_discrete(labels=addline_format(unique(full.dat$author)))

ggplot(data=full.dat,aes(x=author,y=Firmicutes,fill=status)) +
  geom_violin() + 
  geom_jitter() +
  scale_x_discrete(labels=addline_format(unique(alpha_dat$author)))

full.dat %>% filter(author=="Weis et al., 2019") %>% dplyr::select(status,Bacteroidetes) %>% group_by(status) %>%summarize(no_bacteroidetes = sum(Bacteroidetes==0))

```


```{r}
full.dat %>% group_by(author,status) %>% summarise_if(is.numeric,funs(sum)) %>% data.table::data.table()
```

## Permutation based differential abundance
```{r}
otu <- full.dat %>% dplyr::select(-c(1:4))
meta <- full.dat %>% dplyr::select(c(1:4))

group <- meta$status
strata <- meta$author
weights <- meta$N


cores <- detectCores()-2
cl <- makeCluster(cores)
clusterEvalQ(cl, library("permute"))

perm.p <- parApply(cl=cl,otu,2,perm_DA,group=group,type="mean",n.perm=99999,strata=strata,plot=F)
stop_for_problems(cl)

data.frame("p.value"=perm.p,"FDR"=p.adjust(perm.p,method="fdr"))

```


```{r}

ggplot(data=full.dat,aes(x=author,y=Verrucomicrobia,fill=status)) +
  geom_boxplot() +
  scale_y_log10() +
  scale_x_discrete(labels=addline_format(unique(alpha_dat$author)))

```

## Stan - negative binomial
```{r}


stan.dat <- full.dat %>%
            mutate(status = case_when(status=="HC"~0,
                                      status=="PD"~1)) %>%
            arrange(status)


otu <- stan.dat %>% dplyr::select(-c(1:4))
meta <- stan.dat %>% dplyr::select(c(1:4))

colnames(otu)

y <- full.dat %>% dplyr::select(Verrucomicrobia) %>% pull
N <- length(y)
x <- stan.dat %>% dplyr::select(status) %>% pull

groups <- list("starts" = c(min(which(x==0)),
                          min(which(x==1))),
               "ends" = c(max(which(x==0)),
                          max(which(x==1))))

library(rstan)

sm <- stan_model("stan_model.stan")

m <- sampling(sm, data = list("datamatrix" = otu+1,
                              "nreps" = nrow(otu),
                              "notus" = ncol(otu),
                              "N"=2,
                              "start"=groups$starts,
                              "end" =groups$ends),
  pars = c("pi"),
  seed = 1406,
  iter = 500, chains = 4)


saveRDS(m,"stan_DM_phylum.RDS")
shinystan::launch_shinystan(m)


rstan::traceplot(m, inc_warmup = FALSE,nrow=4)
est.pi <- extract(m,"pi")


str(est.pi)
plot(density(est.pi$pi[,1,1]))
plot(density(est.pi$pi[,2,1]))

est <- est.pi$pi[,1,] - est.pi$pi[,2,]



cbind(t(apply(est,2,quantile)),
"mean"=apply(est,2,mean))

```

